### ISSUE TITLE
Evenly spread axis labels

### ISSUE THREAD

**Issue by git-anurag-hub** [recharts/recharts#3305]
- [ ] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
Currently, We are trying to evenly distance axis labels so that our UI doesn't look congested. We want to have a easier way to do this as this can have many other implications. As we are heavily based on recharts we are willing to contribute in the feature building if required.

### What does the proposed API look like?
There can be a prop/attributes which can be passed to X-axis component to evenly spread the axis labels.

**Comment by ckifer** [comment]
Hi @git-anurag-hub 

Can you please give an example of what you would like vs what you have now?

**Comment by git-anurag-hub** [comment]
Hey @ckifer, 
So this is the behaviour right now I am getting: 
<img width="792" alt="Screenshot 2023-02-02 at 1 22 47 PM" src="https://user-images.githubusercontent.com/58766532/216264697-24365768-a433-4d75-a509-92ff76b52db9.png">

This is the behaviour I want: 
 <img width="792" alt="Screenshot 2023-02-02 at 1 28 50 PM" src="https://user-images.githubusercontent.com/58766532/216265223-24587b80-817a-447d-8b21-fdc2560a6cbe.png">
 
 Right now what I am doing is initialising the new variable `interval` to calculate what will be the best interval value for my x-axis in the graph & then passing it as a prop `interval` to X-axis.
```
	const interval = data.length > 10 ? Math.floor(data.length / 10) : 0;
        .
        .
        <XAxis
	        dataKey="name"
		stroke="white"
		opacity={0.4}
	        interval={interval}
	/>
``` 

This works in work cases my fails sometimes. What I think can happen is that all the labels are spread uniformly with the graph in the best possible way by default (or by user choice).

**Comment by ckifer** [comment]
Do you mind throwing your example into a codesandbox with a dummy similar dataset? 

I can't promise anything on this since we're trying to clean the repo up a bit currently, but any info helps

**Comment by git-anurag-hub** [comment]
Sure!
You can find the implementation here: https://codesandbox.io/s/dazzling-kowalevski-8p341n

**Comment by Yilun-Sun** [comment]
Hi @git-anurag-hub 
You can consider add a `tickFormatter`:
```Javascript
tickFormatter={value => `${value}                                      `}
```
And you will get
![image](https://user-images.githubusercontent.com/43896664/217131634-f07dd118-9ec8-44b3-afa9-8b7c3c0903a7.png)

I know there must be a better approach, will look into this soon.

**Comment by git-anurag-hub** [comment]
> You can consider add a `tickFormatter`:

Thanks for the suggestion @Yilun-Sun. But I don't think this will help us as we are not trying to format the tick but to format the axis based on number of ticks.

**Comment by ckifer** [comment]
An actual fix for this issue won't happen to a while since we're trying to focus on making the repo easier to contribute to / fixing docs.

@Yilun-Sun thanks for the help here!

**Comment by nikolasrieble** [comment]
### TLDR
Hardcode which ticks to set using the interval property on the xAxis. https://recharts.org/en-US/api/XAxis#interval

### How you could solve this

So, we happened to recently solve this exact issue with a rather complex workaround - which I do not recommend, but still want to share. 

Recharts does position ticks equidistantly. But then it also disappears all ticks that touch other ticks - that is why larger ticks will disappear. 

If you set an interval value on the xAxis, you force setting ticks, even if they touch each other / i.e. tick text is overlapping. 

We do the following:
- Measure xAxis width (width of charts minus some margins and yAxisWidth)
- Generate ticks from values (i.e. an array of strings)
- Identify the interval value with which all ticks shown would not overlap
   - Measure the tick label width for every string https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/measureText
   - Find the interval={X} value, for which all tick labels do not touch each other 

Note that this becomes a little more complicated when you take tick rotation into account. 

Happy to share more code some time, but preparing this as a comprehensible examples takes some time. In the meantime, take these thoughts. I hope the help.

## How recharts should behave instead

The default recharts tick positioning is not ideal. Equidistant ticks are expected, and aesthetically pleasing. I consider this a feature request and would myself be very happy if we did this better.

**Comment by ckifer** [comment]
Thanks for the info @nikolasrieble - I personally haven't looked into tick calculation at all and therefore this is great info. 

I agree with you at the default I think though id love to POC the different between the two. Definitely a breaking change/3.0 request

**Comment by nikolasrieble** [comment]
In the meantime, have a look at this specific story for time series. It might already solve your need. https://github.com/recharts/recharts/pull/3360

**Comment by nikolasrieble** [comment]
This does not need to be a breaking change. I describe a way to implement this as an enhancement here https://github.com/recharts/recharts/issues/2562

**Comment by christopherkindl** [comment]
Hi @git-anurag-hub, @nikolasrieble,

cc: @severinlandolt

would be great to have option `equidistant` for the prop `interval`. Option `equidistant` would mean that it optimizes always for evenly spaced ticks:

- Both "neighbors" are taken out when labels get dense (e.g. ~Mar 23~ April 23 ~May 23~)
- Always optimizes for equal distance between ticks and not labels (see attached image below)

**Ant Design Charts as reference example:** 

Ant Charts has nailed this problem. Check out the example on the playground and see the magic when you resize the window: https://charts.ant.design/en/examples/line/basic/#line (Update it with my example code below to see the different scenarios)

```
import React, { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';
import { Line } from '@ant-design/plots';

const DemoLine = () => {

const chartData = [
  {
    date: 'Jan 22',
    SemiAnalysis: 2890,
  },
  {
    date: 'Feb 22',
    SemiAnalysis: 2756,
  },
  {
    date: 'Mar 22',
    SemiAnalysis: 3322,
  },
  {
    date: 'Apr 22',
    SemiAnalysis: 3470,
  },
  {
    date: 'May 22',
    // replace with "date: 'This is a long label'" to test for edge case
    SemiAnalysis: 3475,
  },
  {
    date: 'Jun 22',
    SemiAnalysis: 3129,
  },
  {
    date: 'Jul 22',
    SemiAnalysis: 1129,
  },
  {
    date: 'Aug 22',
    SemiAnalysis: 2329,
  },
];
  const config = {
    data: chartData,
    padding: 'auto',
    xField: 'date',
    yField: 'SemiAnalysis',
    xAxis: {
      // type: 'timeCat',
      tickCount: 10,
    },
  };

  return <Line {...config} />;
};

ReactDOM.render(<DemoLine />, document.getElementById('container'));

``` 
![_github](https://github.com/recharts/recharts/assets/53372002/64726c7e-0673-47ff-baf5-cdd6c6de7f1b)

**Comment by ckifer** [comment]
`equidistantPreserveStart` enhanced in 2.9

**Comment by nikolasrieble** [comment]
@git-anurag-hub Can you confirm that the new implementation of `equidistantPreserveStart` solves your requirements?

**Comment by ptakizawa-daitan** [comment]
Hey, I am facing a similar issue on my project. None of the interval options are working for me because I need to have the labels on X Axis to be equidistant and to preserve start and end. Including a `equidistantPreserveStartEnd` option would solve my issue.

**Comment by Zefty** [comment]
@nikolasrieble I've tried your story book implementation and it does seem to be able to make the xaxis dates appear equidistant, however, it does not line up with the graph. See my screenshot below.

You can see that the date Fri 05 seems to be a data point behind compared to on the graph. Any tips on fixing this would be great! 
![Image](https://github.com/user-attachments/assets/546e5489-93b8-4e4e-8127-6924f5e99e8e)

**Comment by ckifer** [comment]
@Zefty That is because of the implementation. This chart is numeric rather than categorical. Unless your tick values are exactly the values you have on the chart, they won't line up because the first tick `Fri 05` is at 00:00, or the start of the day. Your point is halfway through that day.

Changing the chart to a type of category will align the dates with their values on their chart, but it will remove some of the time specific functionality you get with the above.
